rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+---
+feature: intensional_store
+start-date: 2017-08-11
+author: Wout.Mertens@gmail.com
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: Shea Levy, Vladimír Čunát, Eelco Dolstra, Nicolas B. Pierron
+shepherd-leader: Shea Levy
+related-issues: (will contain links to implementation PRs)
+---
+
+# Intensional Store
+
+## TODO / to explain
+
+- flesh out Trust DB locations and updating/querying/merging multiple
+- query service
+- efficient distribution of mappings
+- explain the benefits of late binding and how it improves installs on low-power systems
+- script that uses nix make-content-addressable in /var/lib/nix to generate a store
+- how binary caches provide \$cas and mappings
+
+## Summary
+
+This RFC builds on the implementation of RFC 62 to maximize its benefits.
+
+- Decouple metadata (Trust DB) from the Store, keep per-user
+- Move the Store to `/var/lib/nix`
+- Content checksums include the runtime dependencies
+- Store can be shared read-write on a network share
+- `nix-daemon` becomes optional
+- Store paths optionally provide dependency list in-band
+
+### Benefits
+
+By using content hashes instead of output hashes, we can:
+
+- optimize resource usage
+- reduce binary substitution trust to a single lookup
+- make the Nix store network-writeable and world-shareable
+- predefine mappings from output to content hash without building
+- store paths can be verified without access to the Nix Store DB
+- public trust mappings allow detecting non-reproducible builds
+- easily switch between single- and multi-user setup
+
+Additionally, this is an opportunity to move the Nix store to a filesystem location supported by most non-NixOS systems, namely `/var/lib/nix`.
+
+By "cleaning up" the filesystem state of Nix, a host of possibilities emerge:
+
+- Boot a cloud VM to a specific system by passing a `$cas` name for stage2. The stage1 will auto-download the stage2 if it's missing and switch to it.
+- Cross-compiling can generate `$cas` entries that are reused for native compiles via `$out` mapping. This is useful on low-resource platforms.
+- The Nix store doesn't require any support or metadata. On embedded systems, all management of the store can be performed outside the system.
+- References to `$cas` entries, such as profiles, are no longer tied to a single system.
+- A FUSE filesystem could auto-install `$cas` entries as they are referenced, hanging the I/O until the entry is downloaded and verified.
+- You can copy a store from some other install, and immediately use profiles without having their metadata.
+- Different Nix tooling and metadata implementations can use the same store
+
+… and so on. Decouplying systems brings exponential possibilities.
+
+### Drawbacks
+
+There are some small drawbacks:
+
+- We have to assume that we can always create a working `$cas` derivation, even if the build stores the build path in an opaque way. Otherwise, their build path would have to be a symlink to the `$cas` entry.
```

Comment:
This actually seems like a really nice "feature". Right now this is one of my biggest concerns, how do we handle references that aren't "plaintext"? If the fallback is that there is a symlink it would make this change much less risky. You would get a different CAS path every time the definition changes but at least everything continues to work.

I think the downside is that we need to ensure that the symlink exists wherever this CAS path is used in case it is needed. We would need some way of preserving this across copies and binary caches. (Actually we only need to preserve it if there is exactly 1 symlink referencing it. Once we have two symlinks we can be sure that the CAS path doesn't reference the input path.)

Reply:
> You would get a different CAS path every time the definition changes but at least everything continues to work.

Also of note is that we can write something like a extra derivation to get a more "stable" path from the above by doing the binary replacement inside the derivation.